docs/guides/add-onboarding-flow.mdx

```diff
@@ -0,0 +1,232 @@
+    }
+
+    // User is logged in and the route is protected - let them view.
+    if (userId && !auth.isPublicRoute) return NextResponse.next();
+
+    // If the route is public, anyone can view it.
+    if (auth.isPublicRoute) return NextResponse.next();
+  },
+});
+
+export const config = {
+  matcher: ["/((?!.+\\.[\\w]+$|_next).*)", "/", "/(api|trpc)(.*)"],
+};
+```
+
+Next, create a `layout.tsx` file in `src/app/onboarding` and add the following code to the file. This logic could go in the Middleware, but by adding to the `layout.tsx` to the route the logic remains in one place. This file can also be expanded to handle multiple steps, if multiple steps are required for an onboarding flow.
+
+```tsx
+import { auth } from "@clerk/nextjs/server";
+import { redirect } from "next/navigation";
+
+export default function RootLayout({
+  children,
+}: {
+  children: React.ReactNode;
+}) {
+  // Check if a user has completed onboarding
+  // If yes, redirect them to /dashboard
+  if (auth().sessionClaims?.metadata.onboardingComplete === true) {
+    redirect("/dashboard");
+  }
+
+  return <>{children}</>;
+}
+```
+
+Now that we have the logic for where to direct the user, we’ll need a way to track their onboarding status and note it on their session, let’s dig into that now!
+
+## Update publicMetadata based on onboarding state
+
+Updating a user's publicMetadata as they progress through the flow will allow us to recognize when they have successfully completed their onboarding and, per the logic above, are now able to access the application. [Learn More](https://clerk.com/docs/users/metadata)
+
+To do this you need:
+
+- A method in your backend to securely update the user publicMetadata
+- A process in your frontend with logic to collect and submit all the information for onboarding. In this guide you’ll use an example form.
+
+## Add userUpdate method to your backend
+
+First, add a method in your backend, that will be called on form submission and update the user’s `publicMetadata` accordingly. The example below uses the `clerkClient` wrapper to interact with the Backend API.
+
+Under src/app/onboarding/\_actions.ts add the following code snippet:
+
+```tsx
+"use server";
+
+import { auth, clerkClient } from "@clerk/nextjs/server";
+
+export const completeOnboarding = async (formData: FormData) => {
+  const { userId } = auth();
+
+  if (!userId) {
+    return { message: "No Logged In User" };
+  }
+
+  try {
+    const res = await clerkClient.users.updateUser(userId, {
+      publicMetadata: {
+        onboardingComplete: true,
+        applicationName: formData.get("applicationName"),
+        applicationType: formData.get("applicationType"),
+      },
+    });
+    return { message: res.publicMetadata };
+  } catch (err) {
+    return { error: "There was an error updating the user metadata." };
+  }
+};
+```
+
+Now that we have a method to securely update our user’s `publicMetadata` we can call this server action from a client side form.
```

Comment:
I remove this text that's at the end of every step because it's redundant and informal. The user can simply move on to the next step to see what they need to do next. Maybe in blogs this is okay, because blogs are more informal and the writer speaks directly to the user more, but in technical writing, this is not proper prose.